hail/src/main/scala/is/hail/expr/LowerArrayToStream.scala

```diff
@@ -0,0 +1,65 @@
+package is.hail.expr.ir
+
+import is.hail.expr.types.virtual.{TArray, TContainer, TIterable, TStream}
+
+object LowerArrayToStream {
+  private def nonStreamIterable(node: IR) = node.typ.isInstanceOf[TIterable] && !node.typ.isInstanceOf[TStream]
+
+  private def boundary(node: IR): IR = {
+    val streamified = streamify(node)
+
+    val wrapped = if (streamified.typ.isInstanceOf[TStream] && node.typ.isInstanceOf[TArray])
+      ToArray(streamified)
+    else
+      streamified
+
+    assert(wrapped.typ == node.typ)
+    wrapped
+  }
+
+  private def toStream(node: IR): IR = {
+    val streamified = streamify(node)
+
+    if (nonStreamIterable(streamified))
+      ToStream(streamified)
```

Comment:
Good that this passes, happy to remove the ToArray case. 

I think we still have an issue with a mismatch between what ToStream wraps (any TIterable, which is correct according to our type system), and what boundary handles. As written, in boundary we condition on node.typ.isInstanceOf[TArray], but will miss any other TIterable that we wrapped. It looks like we don't have a test that provides the right conditions to catch that, or I'm not seeing why this condition is impossible.

Reply:
I don't think that can ever happen. When would we ever have an IR of type `TSet` that, when we call `streamify` on it (as it exists in this branch) returns an IR of type TStream?

Reply:
Something that is a child of a node that calls toStream on one of its children?

If I have ToStream(something_that_is_tset), I should get TStream(that_things_elementType)

Reply:
> Something that is a child of a node that calls toStream on one of its children?

boundary calls streamify, not toStream. I still can't think of an example where calling `streamify` on a node of type TSet can give you a TStream.

Reply:
If you call boundary on something whose type is determined transitively, like ArrayMap, whose child determines the type, meaning that node doesn't need to be wrapped directly. So it could be ToStream(a_TIterable) I think, we see TStream, call TArray, and screw it up.

Our rule for wrapping something in ToStream depends on TIterable, on TArray, which means that our boundary rule could be problematic whenever we run into boundary being called on something that is ToArray (I think).

Reply:
if I call `boundary` on an array map, I'm going to get back a TArray. That's the right thing.

We don't wrap in ToStream anywhere in this design except in `toStream`

Reply:
boundary of a ToArray node is going to return ToArray, which is right

Reply:
the initial design from my earlier PR messed this up by wrapping containers in ToStream in the base case of streamify. That certainly caused problems. I think I've resolved them here

Reply:
> if I call boundary on an array map, I'm going to get back a TArray. That's the right thing.

Maybe not the best transitive type example, ArrayFilter works I think. In InferType:

```scala
case ArrayFilter(a, name, cond) =>
        a.typ
```

> We don't wrap in ToStream anywhere in this design except in toStream

Right, we wrap a TIterable, not a TArray (but check only TArray in boundary). I don't think it's sufficient to wrap only TArray (and also that enforces a restriction on our type system).

Reply:
ArrayFilter doesn't work because TypeCheck checks TStreamable. 

I'll look around for an example.

However, again it's not so much about whether we can see this. Our type system allows us to wrap any TIterable in ToStream. If that's the case, we should also, in boundary, be able to properly handle unwrapping that ToStream.

Reply:
I don't think it's possible to construct an IR that fails the assertion in boundary currently. If we create a node that would fail that assertion, we'll change the boundary code to handle whatever has changed

Reply:
basically all of our "stream" nodes (things that can have type tstream with special stream emitters) are either arrays or streams. So the boundary check is currently correct. If we created like a `SetFilter` or something that used streams, we'd have to change it.

Reply:
I see. This seems more fragile than handling the set and dict cases that are made possible by wrapping TIterable in ToStream (in def toStream), but if that's the preference, no problem.

Reply:
I don't understand what these cases are -- I don't see any in our current IR.